torch/_subclasses/fake_tensor.py

```diff
@@ -869,23 +877,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
 
     # No-op if FakeTensorMode is already in use
     def __enter__(self):
+        prev_lift_then_h2d = None
+        if self.avoid_device_init:
+            # See NOTE: [torch.tensor, lift_fresh, and device movement]
+            prev_lift_then_h2d = torch._C._get_lift_then_h2d()
+            torch._C._set_lift_then_h2d(True)
         maybe_prev_fake_mode = torch._C._unset_dispatch_mode(self._mode_key)
         if self is not maybe_prev_fake_mode:
-            self.enter_stack.append((True, maybe_prev_fake_mode))
+            self.enter_stack.append((True, maybe_prev_fake_mode, prev_lift_then_h2d))
```

Comment:
Storing the config directly on FakeTensorMode does give us an annoying amount of flexibility when handling nested modes.

WDYT of just having the config be global state? Downside: you get weird behavior if you change the global state mid-way through the usage of your FakeTensorMode. Maybe that's a reason to stick with setting the config on FakeTensorMode directly, although the expected usage of the config is probably to globally enable it during all of export / model init.

Reply:
What's the problem with nested modes? We set and reset the state correctly on `__enter__/__exit__`.

Reply:
I think I'm mostly just commenting on the fact that it feels unusual to do something like this:
```
with FakeTensorMode(prev_lift_then_h2d=True):
    foo()
    with FakeTensorMode(prev_lift_then_h2d=False):
        bar()
```

But we are giving users the full flexibility to let people do this, and we need to handle it via the extra state reseting on enter/exit.

I agree that this extra state resetting isn't actually that complicated. So if you want to keep the state on the mode then sgtm, I just figured I would call it out.

Reply:
See my other comment. Can we avoid adding this as a ctor and detect when it is necessary if cuda device is used and device is not avialable ?